### ISSUE TITLE
2d slice on a 3d partitioned coarray (derived types)

### ISSUE THREAD

**Issue by neok-m4700** [sourceryinstitute/OpenCoarrays#493]
[issue stat img]: https://img.shields.io/issuestats/i/github/sourceryinstitute/OpenCoarrays.svg?style=flat-square

| Avg response time              |
|:------------------------------:|
| ![Issue Stats][issue stat img] |

# Defect/Bug Report #

- OpenCoarrays Version: `4cc41bf4f301190c (1.9.3+)`
- Fortran Compiler: `GNU Fortran (GCC) 7.3.0 (release)`
- C compiler used for building lib: `gcc (GCC) 7.3.0 (release)`
- Installation method: `cmake ..; make -j; make install`
- Output of `uname -a`: `Linux <...> 4.13.0-32-generic #35-Ubuntu SMP Thu Jan 25 09:13:46 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
- MPI library being used: `mpich 3.2.1 (release)`
- Also tested with `OpenMPI 3.0.0  (release)`
- Machine architecture and number of physical cores: `x86_64 16 cores`
- Version of CMake: `cmake version 3.10.2 (release)`

## Observed Behavior ##
Plane slicing a 3d array on certain directions fails with error message
`Fortran runtime error on image <...>: libcaf_mpi::caf_get_by_ref(): rank out of range.`

## Steps to Reproduce ##

**slice.f90**
```fortran
program slice
   type coarr
      real, allocatable :: a(:, :, :)[:, :, :]
   end type

   type(coarr) :: co
   integer :: nimg, me, z, nx, ny, nz, north, south, mex, mey, mez, coords(3)
   real, allocatable :: bufxz(:, :) ! a plane (2d) slice, normal in the y direction

   nx = 6
   ny = 4
   nz = 2

   me = this_image()
   nimg = num_images()

   if (nimg /= 8) stop

   allocate(co % a(nx, ny, nz)[1:2, 1:2, *])
   allocate(bufxz(nx, nz))

   co % a = reshape([(z, z=1, nx * ny * nz)], shape(co % a))

   coords = this_image(co % a)
   mex = coords(1)
   mey = coords(2)
   mez = coords(3)

   north = mey + 1
   south = mey - 1

   sync all
   if (north <= 2) then
      z = image_index(co % a, [mex, north, mez])
      sync images(z)
      bufxz = co % a(1:nx, 1, 1:nz)[mex, north, mez]
      co % a(1:nx, ny, 1:nz) = bufxz
   end if
   if (south >= 1) then
      z = image_index(co % a, [mex, south, mez])
      sync images(z)
      bufxz = co % a(1:nx, 1, 1:nz)[mex, south, mez]
      co % a(1:nx, ny, 1:nz) = bufxz
   end if
   sync all

   deallocate(co % a, bufxz)
end program
```
**build_run**
```bash
#!/usr/bin/env bash
caf -o slice slice.f90
cafrun -np 8 ./slice
```

# Proposed solution #
In the previous example, `dst_cur_dim` is incorrectly incremented when transitioning from `dim=1` to `dim=2` with `riter->u.a.mode[1] == CAF_ARR_REF_SINGLE` (singleton), leading to a destination rank failure on assignement to the 2d plane `bufxz`.

**issue_493.patch**
```diff
--- src/mpi/mpi_caf.c  2018-01-29 17:38:05.000000000 +0100
+++ src/mpi/mpi_caf.c  2018-02-08 00:18:21.372970368 +0100
@@ -4624,8 +4624,9 @@
                         }
                     }
 
-                  /* Only increase the dim counter, when in an array ref.  */
-                  if (in_array_ref && dst_cur_dim < dst_rank)
+                  /* Only increase the dim counter, when in an array ref, and
+                     MODE != CAF_ARR_REF_SINGLE (delta == 1) see caf_array_ref_t. */
+                  if (in_array_ref && dst_cur_dim < dst_rank && delta != 1)
                     ++dst_cur_dim;
                 }
               size *= (ptrdiff_t)delta;
@@ -4772,8 +4773,9 @@
                           dst->dim[dst_cur_dim]._stride = size;
                         }
                     }
-                  /* Only increase the dim counter, when in an array ref.  */
-                  if (in_array_ref && dst_cur_dim < dst_rank)
+                  /* Only increase the dim counter, when in an array ref, and
+                     MODE != CAF_ARR_REF_SINGLE (delta == 1) see caf_array_ref_t. */
+                  if (in_array_ref && dst_cur_dim < dst_rank && delta != 1)
                     ++dst_cur_dim;
                 }
               size *= (ptrdiff_t)delta;
```
[links]:#
[Google]: https://www.google.com
[Stack Overflow]: https://stackoverflow.com
[fenced codeblocks]: https://help.github.com/articles/creating-and-highlighting-code-blocks/
[GFortran bug tracker]: https://gcc.gnu.org/bugzilla/
[Intel MPI]: https://software.intel.com/en-us/intel-mpi-library
[MPT]: http://www.sgi.com/products/software/sps.html
[MVAPICH]: http://mvapich.cse.ohio-state.edu
[MPICH]: https://www.mpich.org
[Open-MPI]: https://www.open-mpi.org
[CONTRIBUTING.md]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md
[Helping Out]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md#helping-out
[official release]: https://github.com/sourceryinstitute/OpenCoarrays/releases
[CMake install]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/INSTALL.md#cmake-scripts
[homebrew]: https://brew.sh
[issues page]: https://github.com/sourceryinstitute/OpenCoarrays/issues
[mailing list]: https://groups.google.com/forum/#!forum/opencoarrays
[install.sh]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/install.sh

**Comment by rouson** [comment]
@neok-m4700 Thanks for this contribution.   Please fork the repository and submit a pull request with the change.  If you mention issue #493 in the pull request, that should hyperlink it to this issue.    Please see the [CONTRIBUTING.md](https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md) document. One advantage of submitting a pull requests is that pull requests automatically generate testing of the patch with GCC 6 and 7 on Linux and macOS.  The test results will be helpful to those reviewing the pull request.

**Comment by zbeekman** [comment]
Was fixed in #495 

Tests will be merged in PR #500
